File: MediafyDemoAppObjectiveC/MediafyDemoAppObjectiveC/IntegrationCases/SDK/MediafyNativeViewController.m
/*
 * Mediafy SDK
 *
 * Copyright 2024 Mars Technologies Ltd. All rights reserved.
 */

#import "MediafyNativeViewController.h"

@interface MediafyNativeViewController ()

@property (nonatomic) MediafyNativeAdUnit * nativeAdUnit;
@property (nonatomic, nullable) MediafyNativeAd * nativeAd;

@end

@implementation MediafyNativeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self loadAd];
}

- (void)loadAd {
    // 1. Create a MediafyNativeAdUnit
    self.nativeAdUnit = [[MediafyNativeAdUnit alloc] init];
    
    // 2. Configure the MediafyNativeAdUnit 
    MediafyNativeAssetData *cta = [
        [MediafyNativeAssetData alloc]
        initWithType:MediafyDataAssetCtatext
    ];
    
    cta.length = @15;
    
    MediafyNativeAssetTitle *title = [
        [MediafyNativeAssetTitle alloc]
        initWithLength:90
    ];
    
    title.required = @YES;
    
    MediafyNativeAssetImage *icon = [
        [MediafyNativeAssetImage alloc]
        initWithMinimumWidth:20
        minimumHeight:20
    ];
    
    icon.type = MediafyImageAsset.Icon;
    
    MediafyNativeAssetImage *image = [[MediafyNativeAssetImage alloc] init];
    
    image.required = @YES;
    image.width = @1200;
    image.height = @627;
    
    image.type = MediafyImageAsset.Main;
    
    MediafyNativeAssetData * description = [
        [MediafyNativeAssetData alloc]
        initWithType:MediafyDataAssetDescription
    ];
    
    description.required = @YES;
    description.length = @150;
    
    MediafyNativeAssetData * rating = [
        [MediafyNativeAssetData alloc]
        initWithType:MediafyDataAssetRating
    ];
    
    
    MediafyNativeParameters * parameters = [MediafyNativeParameters new];
    parameters.assets = @[cta, title, icon, image, rating, description];
    
    MediafyNativeEventTracker * eventTracker = [
        [MediafyNativeEventTracker alloc]
        initWithEvent:MediafyEventType.Impression
        methods:@[MediafyEventTracking.Image, MediafyEventTracking.js]
    ];
    
    parameters.eventtrackers = @[eventTracker];
    parameters.context = MediafyContextType.Social;
    parameters.placementType = MediafyPlacementType.FeedContent;
    parameters.contextSubType = MediafyContextSubType.Social;
    
    self.nativeAdUnit.parameters = parameters;
    
    __weak typeof(self) weakSelf = self;
    [self.nativeAdUnit loadAdWithCompletion:^(MediafyNativeAd * _Nullable ad, NSError * _Nullable error) {
        if (error != nil || ad == nil) {
            return;
        }
        
        weakSelf.nativeAd = ad;
        weakSelf.nativeAd.delegate = self;
        
        // 3. Render the native ad
        weakSelf.titleLabel.text = ad.title;
        weakSelf.bodyLabel.text = ad.text;
        weakSelf.sponsoredLabel.text = ad.sponsoredBy;
        
        [weakSelf.mainImageView setImageFromURLString:ad.imageUrl
                                          placeholder:[UIImage systemImageNamed:@"photo.artframe"]];
        [weakSelf.iconView setImageFromURLString:ad.iconUrl
                                          placeholder:[UIImage systemImageNamed:@"photo.artframe"]];
        [weakSelf.callToActionButton setTitle:ad.callToAction forState:UIControlStateNormal];
    }];
}

// MARK: - MediafyNativeAdEventDelegate

- (void)adDidExpireWithAd:(MediafyNativeAd *)ad {
    // Called when the ad expired
}

- (void)adWasClickedWithAd:(MediafyNativeAd *)ad {
    // Called when the ad was clicked
}

- (void)adDidLogImpressionWithAd:(MediafyNativeAd *)ad {
    // Called when the impression was logged
}

@end

File: MediafyDemoAppSwift/MediafyDemoAppSwift/IntegrationCases/SDK/MediafyNativeViewController.swift
/*
 * Mediafy SDK
 *
 * Copyright 2024 Mars Technologies Ltd. All rights reserved.
 */

import UIKit
import MediafySDK

final class MediafyNativeViewController: NativeBaseViewController {
    
    private var nativeAdUnit: MediafyNativeAdUnit!
    private var nativeAd: MediafyNativeAd?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        loadAd()
    }
    
    private func loadAd() {
        // 1. Create a MediafyNativeAdUnit
        nativeAdUnit = MediafyNativeAdUnit()
        
        // 2. Configure the MediafyNativeAdUnit
        let cta = MediafyNativeAssetData(type: .ctatext)
        cta.length = 15
        
        let title = MediafyNativeAssetTitle(length: 90)
        title.required = true
        
        let icon = MediafyNativeAssetImage(minimumWidth: 50, minimumHeight: 50)
        icon.type = .Icon
        
        let image = MediafyNativeAssetImage()
        image.required = true
        image.width = 1200
        image.height = 627
        image.type = .Main
        
        let description = MediafyNativeAssetData(type: .description)
        description.required = true
        description.length = 150
        
        let rating = MediafyNativeAssetData(type: .rating)
        
        let parameters = MediafyNativeParameters()
        parameters.assets = [cta, title, icon, image, rating, description]
        
        let eventTracker = MediafyNativeEventTracker(
            event: .Impression,
            methods: [.Image, .js]
        )
        
        parameters.eventtrackers = [eventTracker]
        parameters.context = .Social
        parameters.placementType = .FeedContent
        parameters.contextSubType = .Social
        
        nativeAdUnit.parameters = parameters
        
        nativeAdUnit.loadAd { [weak self] ad, error in
            guard let self = self else {
                return
            }
            
            guard let ad = ad, error == nil else {
                return
            }
            
            self.nativeAd = ad
            self.nativeAd?.delegate = self
            
            // 3. Render the native ad
            self.titleLabel.text = ad.title
            self.bodyLabel.text = ad.text
            self.sponsoredLabel.text = ad.sponsoredBy
            
            self.mainImageView.setImage(from: ad.imageUrl, placeholder: UIImage(systemName: "photo.artframe"))
            self.iconView.setImage(from: ad.iconUrl, placeholder: UIImage(systemName: "photo.artframe"))
            self.callToActionButton.setTitle(ad.callToAction, for: .normal)
            
            self.nativeAd?.registerView(view: self.view, clickableViews: [self.callToActionButton])
        }
    }
}

// MARK: - MediafyNativeAdEventDelegate

extension MediafyNativeViewController: MediafyNativeAdDelegate {
    
    func adDidExpire(ad: MediafyNativeAd) {
        // Called when the ad expired
    }
    
    func adWasClicked(ad: MediafyNativeAd) {
        // Called when the ad was clicked
    }
    
    func adDidLogImpression(ad: MediafyNativeAd) {
        // Called when the impression was logged
    }
}
